tempesta_fw/sock.c

```diff
@@ -482,15 +485,39 @@ ss_tcp_process_data(struct sock *sk)
 		skb_dst_drop(skb);
 		skb->dev = NULL;
 
-		/* SKB may be freed in processing. Save the flag. */
-		tcp_fin = tcp_hdr(skb)->fin;
+		/*
+		 * When GRO is used, multiple SKBs may be joined in one
+		 * big SKB. These SKBs are attached through frag_list.
+		 * Interpret the big SKB as a set of separate smaller
+		 * SKBs for processing. Move the FIN flag if necessary.
+		 */
+		frag_list = skb_shinfo(skb)->frag_list;
+		if (frag_list) {
+			struct sk_buff *frag_skb, *last_skb = frag_list;
+			skb_walk_frags(skb, frag_skb) {
+				frag_skb->nohdr = 0;
+				ss_skb_adjust_data_len(skb, -frag_skb->len);
+				last_skb = frag_skb;
+			}
```

Comment:
Then ```ss_skb_process()``` shall not walk through the fragments. And this is good to do in this patch to clearly express the logic movement.